ambari-server/src/main/java/org/apache/ambari/server/state/UpgradeHelper.java

```diff
@@ -464,6 +423,56 @@ public UpgradePack suggestUpgradePack(String clusterName,
             continue;
           }
 
+          HostsType hostsType = mhr.getMasterAndHosts(service.serviceName, component);
+
+          // only worry about adding future commands if this is a start/restart task
+          boolean taskIsRestartOrStart = functionName == null || functionName == Type.START
+              || functionName == Type.RESTART;
+
+          // see if this component has an add component task which will indicate
+          // we need to dynamically schedule some more tasks by predicting where
+          // the components will be installed
+          String serviceAndComponentHash = service.serviceName + "/" + component;
+          if (taskIsRestartOrStart && addedComponentsDuringUpgrade.containsKey(serviceAndComponentHash)) {
+            AddComponentTask task = addedComponentsDuringUpgrade.get(serviceAndComponentHash);
+
+            Collection<Host> candidateHosts = MasterHostResolver.getCandidateHosts(cluster,
+                task.hosts, task.hostService, task.hostComponent);
+
+            // if we have candidate hosts, then we can create a structure to be
+            // scheduled
+            if (!candidateHosts.isEmpty()) {
+              if (null == hostsType) {
+                // a null hosts type usually means that the component is not
+                // installed in the cluster - but it's possible that it's going
+                // to be added as part of the upgrade. If this is the case, then
+                // we need to schedule tasks assuming the add works
+                hostsType = HostsType.normal(
+                    candidateHosts.stream().map(host -> host.getHostName()).collect(
+                        Collectors.toCollection(LinkedHashSet::new)));
+              } else {
+                // it's possible that we're adding components that may already
+                // exist in the cluster - in this case, we must append the
+                // structure instead of creating a new one
+                Set<String> hostsForTask = hostsType.getHosts();
```

Comment:
This is really the big change here - we need to still add the candidate hosts to the `hostsType` if the `hostsType` is not null. It would be not null if there are already some components installed of the kind we're adding.